### PR TITLE
Introduce `NewLoggingHTTPTransport` and deprecate `NewTransport`

### PR DESCRIPTION
**Pull request by detro**
Related: https://github.com/hashicorp/terraform-provider-tfe/pull/479
Related: #2 
Closes: #546

## Background

Current SDK offers an helper `http.RoundTripper` in `helpers/logging`, that is created via `logging.NewTransport()`. This `http.RountTripper` can then be used as `.Transport` in an `http.Client`.

When logging level is _debug or higher_, it does a simple dump of the whole content of the request and of the following response.

This exposes any sensitive information that might be going over HTTP, and there is no way for developers using this code, to setup any filtering.

## Proposal

This PR introduced a 2 new implementations of `http.RoundTripper`, one designed to log via the provider root logger, the other via a user-defined subsystem.

```golang
func NewLoggingHTTPTransport(t http.RoundTripper) *loggingHttpTransport {
	return &loggingHttpTransport{"", false, t}
}

func NewSubsystemLoggingHTTPTransport(subsystem string, t http.RoundTripper) *loggingHttpTransport {
	return &loggingHttpTransport{subsystem, true, t}
}
```

This transport can be used exactly like the previous one, but it's built around the `tflog` package of [terraform-plugin-log](https://github.com/hashicorp/terraform-plugin-log), and offers an opt-in hook to configure the HTTP Request `context.Context`.

For example, if the user wants to setup log filtering against the provider root logger, and ensure that it's applied to each request handled by this transport:

```golang
// ctx == context.Context, provided by the SDK...

ctx = tflog.MaskFieldValuesWithFieldKeys(ctx, "MY_SENSITIVE_FIELD")
ctx = tflog.MaskMessageStrings(ctx, "MY_SECRET_STRING")

transport := logging.NewLoggingHTTPTransport(http.DefaultTransport)
client := http.Client{
  Transport: transport,
}

req, _ := http.NewRequest("GET", "https://www.terraform.io", nil)

res, err := client.Do(req.WithContext(ctx))
```

As a consequence, the existing `NewTransport` would be documented as deprecated.

## Related

* How to setup [log filtering](https://www.terraform.io/plugin/log/filtering) with `terraform-plugin-log`
* Fast follow-up: #1007

### CODE REVIEW

File: helper/logging/logging_http_transport.go
Comment (detro) [reply helper/logging/logging_http_transport.go:174]: Will look into it

File: helper/logging/logging_http_transport.go
Comment (detro) [reply helper/logging/logging_http_transport.go:16]: Totally random: I can adopt any format that is good for this context.

I was just looking for a conversation starter.

File: Makefile
Comment (detro) [reply Makefile:35]: Felt like a perfect place for "leaving the campfire better than I found it".

File: helper/logging/logging_http_transport.go
Comment (detro) [reply helper/logging/logging_http_transport.go:60]: In the light of the above, this will be entirely removed.

File: helper/logging/logging_http_transport.go
Comment (detro) [reply helper/logging/logging_http_transport.go:174]: Actually, `ReadMIMEHeader()` doesn't like the full body of a req/res (what I have here), but only the sub-section that contains actual headers.

So I can't use it to read the `HTTP Request Line`, nor the `HTTP Message Status Line`.
I'll see if it makes sense to ue it for the headers though.

File: helper/logging/logging_http_transport.go
Comment (bflad) [reply helper/logging/logging_http_transport.go:16]: This is an excellent conversation starter, then! Maybe we should provide some guidance and document recommendation for field keys (e.g. in the /plugin/log/writing documentation?).

I actually do not mind to much whichever guidance we give, although I will throw it out there that while uppercase field keys can certainly make them easier to spot in amongst the other text, they _may_ come across as overly "aggressive" as in the plaintext version of logs, e.g.

```text
1970-01-01T00:00:00 [DEBUG] Example message: HTTP_OP_TYPE=HTTP_REQ HTTP_REQ_METHOD=GET HTTP_REQ_URI=https://terraform.io HTTP_REQ_VERSION=1.1
```

versus

```text
1970-01-01T00:00:00 [DEBUG] Example message: http_op_type=HTTP_REQ http_req_method=GET http_req_uri=https://terraform.io http_req_version=1.1
```

Just my personal opinion though. 😄

File: helper/logging/logging_http_transport.go
Comment (bflad) [reply helper/logging/logging_http_transport.go:174]: I was peeking at the `ReadResponse` function in https://go.dev/src/net/http/response.go, which looks like it does `ReadLine()` then `ReadMIMEHeaders()`, but then I realized you already have a [`http.Response` object](https://pkg.go.dev/net/http#Response). Can we just use that information? How about with the request?

Comment (bflad) [reply helper/logging/logging_http_transport.go:174]: The usage of `httputil.Dump*` functionality in the original implementation may have been to just simplify the plaintext-only representation there. Since we want to provide richer information, I think we can use the higher-level abstractions available in `http.Request` and `http.Response`. 😄

**Comment by detro** [comment]
I have implemented all the feedback of the PR review so far.

I still need to:

- re-do a test that uses log filtering
- update the PR proposal above with the actual implementation
- website doc

Additionally, I have set explicitly the version of golangci-lint (`1.47.1`), given that trying to get the GH Action to pick automatically the latest version didn't work (caching?).

I can revert this once we are reaching a mergeable state for this PR.

File: helper/logging/logging_http_transport.go
Comment (detro) [reply helper/logging/logging_http_transport.go]: I haven't finished yet, but I think you are right: for `http.Response` shouldn't be too tricky to just use the struct as is.

For `http.Request` I don't know yet, but will give it another go.

The key issue though seems to be the bit that those `Dump*` function did with the body: being the body a `ReadCloser`, it's tricky to grab that without interfering with the request/response. Those `Dump*` utilities do lots of "tricks" around the body to ensure they return a copy, without affecting the underlying buffer and the expectation of http client that is sending/receiving.

I'll have to do the same to avoid disrupting things, and this is the main reason I was sticking with the `Dump*` methods: splitting strings around the most well-formed standard heading in the industry, felt a safe long term bet.

But 🙆‍♂️ .

Comment (detro) [reply helper/logging/logging_http_transport.go]: Update (focusing on the response right now, but request isn't much different).

1. A detail that we missed so far is that `Response.Body` contains the WHOLE body of the response. Meaning, while I can use `Response.Header`, `Response.Status` etc. to populate most fields, `Response.Body` contains the whole body of the request (html/json/whatever), plus all the same headers above it.

2. In addition to that, the process to read the bytes out of a `Response.Body`, while preserving the response to that the client can consume it, it's requiring me to re-implement `httputil.DumpResponse` entirely. So I'm planning to keep using it instead, to grab the `Body`, while using the `Response.Header` (etc) fields to populate the other fields.

What do you advice to do about 1. and what do you think of 2.?

File: helper/logging/logging_http_transport.go
Comment (bflad) [inline helper/logging/logging_http_transport.go:156]: It's a bummer that these headers go beyond just `User-Agent` (which maybe we should set one to mention `terraform-plugin-sdk`, if its not already set?), since that would certainly simplify this logic.

File: helper/logging/logging_http_transport.go
Comment (bflad) [inline helper/logging/logging_http_transport.go:36]: Can you ensure there is a changelog file added for these two new function enhancements and a note about the deprecation of the existing transport? Thanks!

File: helper/logging/logging_http_transport_test.go
Comment (bflad) [inline helper/logging/logging_http_transport_test.go:48]: I wonder if instead of putting the body into the message field, we should instead put it into its own `tf_http_req_body` (or whatever) field and set the message to something like "Sending HTTP request"?

File: helper/logging/logging_http_transport_test.go
Comment (bflad) [inline helper/logging/logging_http_transport_test.go:79]: Similar note about potentially using a new `tf_http_res_body` field and setting message to "Received HTTP response" so its clear what folks are seeing in the logs.

**Review by bflad — COMMENTED** [review]
Overall I think this is almost there -- just a few little things then should be good to go. 👍

File: helper/logging/logging_http_transport.go
Comment (detro) [reply helper/logging/logging_http_transport.go:94]: I was conflicted if to do that or not. 
I didn't want to have a case where user sets the subsystem name to `""` and get an unexpected fall back (though, it's a heck of a corner case).

I'll refactor and add an error or something.

File: helper/logging/logging_http_transport.go
Comment (detro) [reply helper/logging/logging_http_transport.go:36]: yep, now that signatures are agreed, I can do that.

File: helper/logging/logging_http_transport.go
Comment (detro) [reply helper/logging/logging_http_transport.go:189]: Let me see if I can refactor this way. Would certainly simplify things.

File: helper/logging/logging_http_transport_test.go
Comment (detro) [reply helper/logging/logging_http_transport_test.go:48]: I think we talked in zoom briefly about this. The only argument I could find for this, was that the "log masking" can be applied to a portion of the `@message` this way.
If we make the body a field, it's a "all or nothing": either they mask the whole value, or not.

I can go either way. I just feel that, in the case of masking, it would be "too blunt". 

What's your preference?

File: helper/logging/logging_http_transport_test.go
Comment (detro) [reply helper/logging/logging_http_transport_test.go:79]: As above.
I can go this way, I don't mind. Just wanted a final check about the log masking limitations we would encounter.

File: helper/logging/logging_http_transport.go
Comment (detro) [reply helper/logging/logging_http_transport.go:156]: Yeah, that's why this `http.DumpRequestOut` utility is so tricky, because internally it essentially "pretends" to send the request, and grabs the output at the "end of the pipe" (if it makes sense).

Comment (detro) [reply helper/logging/logging_http_transport.go:156]: But at least you made me get rid of manual parsing and relying on `textproto`, and that's good.

File: helper/logging/logging_http_transport.go
Comment (detro) [reply helper/logging/logging_http_transport.go:189]: Yeah, I think this will work. Thank you.

File: helper/logging/logging_http_transport.go
Comment (detro) [reply helper/logging/logging_http_transport.go:94]: I went for a godoc comment that warns of the "fallback" behaviour.

File: helper/logging/logging_http_transport_test.go
Comment (bflad) [reply helper/logging/logging_http_transport_test.go:48]: Aye, yes given the current filtering functionality, that does make this a little less developer friendly. 🙁 Maybe it is worth getting this developer-facing problem sorted before potentially changing it later? Flipping this problem around, being able to also have the option to redact the entire body via field key would be kinda nice (and is generally a feature for HTTP loggers)!

Regardless of where we land with this particular functionality, I think adding the ability to mask _portions_ of field values makes sense. Whether that masking applies to "any key" or "this specific key" is probably an important distinction, sigh. I'll create an upstream feature request issue for that. If distinguishing between message masking and field masking is too much of an ask, then maybe having a "just mask all the things" thing is appropriate.

File: helper/logging/logging_http_transport.go
Comment (bflad) [inline helper/logging/logging_http_transport.go:203]: Drive-by note: Do we _need_ to copy the bytes here? We are not modifying them and they could potentially be "a lot" of memory. Having a reader pointing at the same bytes we originally read shouldn't have any effect in this case, I think.

Comment (detro) [reply helper/logging/logging_http_transport.go:203]: I don't know: we would have a shared piece of memory between the logging library, and the user of that response body.

And if we make modification to that shared portion of memory, we end up changing it under the hood for the user of the response body too.

File: helper/logging/logging_http_transport_test.go
Comment (detro) [reply helper/logging/logging_http_transport_test.go:48]: I think it's unlikely that something that a user explicitly wants masked in a message, would be ok to leave plan in a field (and viceversa).

I think the correct thing to do would be to apply the same "find and replace" we do to messages, to the content of each key in the fields.

Comment (detro) [reply helper/logging/logging_http_transport_test.go:48]: All right, here is a PR to address the missing feature in terraform-plugin-log: https://github.com/hashicorp/terraform-plugin-log/pull/87

Comment (detro) [reply helper/logging/logging_http_transport_test.go]: Code that implements this, is now pushed.

The version of `terraform-plugin-log` has to be released before we can merge though.

File: helper/logging/logging_http_transport_test.go
Comment (detro) [reply helper/logging/logging_http_transport_test.go]: Done.

File: helper/logging/logging_http_transport.go
Comment (bflad) [inline helper/logging/logging_http_transport.go:117]: Should these messages be exported in the public interface? They're okay to be constants, but I'm not sure we want to signal any sort of compatibility or code-desirable use cases for the string values themselves.

File: helper/logging/logging_http_transport.go
Comment (bflad) [inline helper/logging/logging_http_transport.go:125]: Taking another look at https://github.com/hashicorp/terraform-plugin-sdk/issues/546, I think we may want to consider one more thing there to cover the ask: generating our own "HTTP request ID". Having a separately generated ID means you can peek at a request or response, copy that ID value, then find anything related to that specific HTTP request/response.

This might be something like the `tf_req_id` log field, which is managed upstream by terraform-plugin-go to automatically generate a UUID value that gets persisted with the logger contexts for each RPC request/response. [Upstream code reference](https://github.com/hashicorp/terraform-plugin-go/blob/4a24b5d5377f223ca890cdada0b6fdbab7410d45/internal/logging/context.go#L58-L71).

I think we can separately decide whether we should be setting our own User-Agent if none is configured, so we can add that generated ID into there are well.

File: helper/logging/logging_http_transport.go
Comment (bflad) [reply helper/logging/logging_http_transport.go:228]: Downstream `io.Reader` consumers cannot modify the underlying bytes as far as I know and we control this implementation, but let's add a code comment explaining that choice then, e.g. "we're doing this in case its modified" 👍

**Review by bflad — APPROVED** [review]
Approving to unblock and this looks good overall, however I do think we should do something for #546 specifically (or not close it yet) as part of this. 👍

File: helper/logging/logging_http_transport.go
Comment (detro) [reply helper/logging/logging_http_transport.go:228]: I'll give it a quick test: maybe I have over-thought this and it's way easier than I thought.

File: helper/logging/logging_http_transport.go
Comment (detro) [reply helper/logging/logging_http_transport.go:117]: I was conflicted indeed. Yeah, I'll make them private, good call.

File: helper/logging/logging_http_transport.go
Comment (detro) [reply helper/logging/logging_http_transport.go:125]: I think it makes sense to add that UUID. But are you asking me to use the one set by `terraform-plugin-go` or to add a new UUID, generated for each Req/Res pair that goes through the Transport?

Comment (detro) [reply helper/logging/logging_http_transport.go:125]: To clarify: I'm asking because each request, comes in with a `Context` and we have documented (and will add it to the website doc), that they need to pass in the context they receive as part of the RPC callbacks into the provider code.

So, my assumption is, that UUID is... already there?

Comment (bflad) [reply helper/logging/logging_http_transport.go:125]: Sorry I wasn't clear 😅 I meant generate a new UUID for each HTTP request, separate from the other one. Each RPC may result in multiple HTTP requests so its good to provide separate IDs for each. 👍

Comment (detro) [reply helper/logging/logging_http_transport.go:125]: Ah, great.

Yeah, totally happy to do it, if that closes #546 properly: seems like a long overdue ask 👍

File: helper/logging/logging_http_transport.go
Comment (detro) [reply helper/logging/logging_http_transport.go:228]: Silly me: the string replacement we do during masking is going to generate a new string every time, so there is no chance of us messing up the same underlying byte buffer.

I can entirely remove this copying.

**Comment by detro** [comment]
All right, I think this is addressing everything @bflad - Indeed I'm pretty happy we get to address #546 in its entirety. Also, got rid of the pointless copying around of the response buffer.

Even if you approved, I think I'll wait for a confirmation the UUID looks good before merging.

**Review by bflad — APPROVED** [review]
Looks good to me 🚀

**Pull request merged**
